Saner API for state retrieval

maichmueller opened the pull request:
Hi,

I have been working with the code for the past few days and certain API decisions are problematic at the moment. Consider this example:
```
def get_init_state():
    parser = mi.PDDLParser(path_to_domain, path_to_problem)
    factory = parser.get_pddl_factories()
    domain = parser.get_domain()
    problem = parser.get_problem()
    aag = mi.GroundedApplicableActionGenerator(problem, factory)
    space = mi.StateSpace.create(problem, factory, aag, mi.StateRepository(aag))
    
    state = space.get_initial_state()  # this is not a state
    state = space.get_state(state)  # this is not a state either yet
    state = state.get_state()  # now it is a state
    return state
```
Incidentally, the above function is also currently a segfault trigger, since the `StateVertex` returned by the `StateSpace` does not keep the `StateSpace` alive, while the `State` only keeps the `StateVertex` alive.

Aside from the segfault, similar issues arise when querying for atoms in a state: `state.get_fluent_atoms()` does not, in fact,  return atoms, it returns atom indices which need to be given to fluent atom lists to receive the actual atoms. 

Since the current API is confusing, I would like to change the bindings to reflect what you receive now. This PR addresses this as such:
- (fixes segfault issue)
- a `State` object now returns atom indices from functions `get_fluent_atom_indices` (and likewise named equivalents).
- StateSpace `get_state` --> `get_state_vertex` and likewise for multiple. I also added actual `get_state` functions that prevent the need to go through a vertex.

With my changes, the above snippet would then read:
```
def get_init_state():
    parser = mi.PDDLParser(path_to_domain, path_to_problem)
    factory = parser.get_pddl_factories()
    domain = parser.get_domain()
    problem = parser.get_problem()
    aag = mi.GroundedApplicableActionGenerator(problem, factory)
    space = mi.StateSpace.create(problem, factory, aag, mi.StateRepository(aag))
    
    state_index = space.get_initial_state_index()
    state_vertex = space.get_state_vertex(state_index)
    state = state_vertex.get_state()

# alternatively

    state = space.get_initial_state()  # now returns an actual State object
    return state
```

Let me know what you think about these changes or if you want something changed.

maichmueller commented:
@simon-stahlberg , the segfault should be fixed as well. I added the necessary return policies where possible and enabled manual keep_alive calls between owners and internals.

drexlerd commented:
There should not be segfaults caused by 
```cpp
    const StateVertexList& get_state_vertices() const;
    const StateVertex& get_state_vertex(Index state) const;
```
Perhaps there is just a bug in the current return policy. There is a lot of additional boilerplate code added in this commit, which to me looks redundant. Why should it be needed here and not in all other places? I will investigate...

drexlerd commented:
Here is the error, and it occurs in many places in the bindings! 
```cpp
.def("get_states", &StateSpace::get_states, py::return_value_policy::reference_internal);
```
The get states method returns a `const StateVertexList&` which is not declared as an opaque type and misses a copy as well + an additional keep_alive for the get_states method to ensure that the state space is not destroyed.

Recap: Opaque types ensure that we get the keep alive from accessing elements, see py::bind_vector, copy ensure that we are not overwriting internal c++ memory.

drexlerd commented:
Exactly, the vertex does not keep the state alive but it can be fixed with `reference_internal` because it is a const reference to a custom type. However, The `reference_internal` does not work for the `get_states`. When you access an element and let the vector and the state space go out of scope, you will get a segfault.

Regarding the API change: Notice that the vertex indices do not necessarily match the state indices. Now, if I want to access the successor states of the initial state, I first have to get the vertex index somehow. I don't see how adding these duplicate functions will make things easier. The current API will give you the index to the root vertex in the graph and you can easily access the successor vertices and their underlying states. The core idea of state space is to be a "proxy" to the underlying graph, where the functions just forward under a different name that better matches the context of state spaces. Perhaps it is a bit misleading to use the naming `get_states` instead of `get_state_vertices`, etc. The benefit of this idea is that you can throw the underlying graph into the boost BGL library to compute useful stuff for the state space. 

My suggestion is that we should either only have those proxy functions or we remove them an have functions like the one that you suggested but having both is very confusing to me. In the latter case, it requires writing additional iterators that yield states instead of state vertices, which will add some additional boilerplate.

The other boilerplate I was referring to is in `bindings.cpp`: e.g. `insert_into_list` it looks unnessary given that one can fix the keep-alive chain through opaqueness as it is done in other cases already.

drexlerd commented:
Alternatively, we could move the existing bindings into a state graph class and provide your suggested bindings directly in the state space. However, the issue with accessing the vertex index from a given state still remains.

maichmueller commented:
> Regarding the API change: Notice that the vertex indices do not necessarily match the state indices. Now, if I want to access the successor states of the initial state, I first have to get the vertex index somehow. I don't see how adding these duplicate functions will make things easier. The current API will give you the index to the root vertex in the graph and you can easily access the successor vertices and their underlying states.

You can perhaps easily access the successor indices, but the name `state` is too overloaded at the moment. Users of the lib cannot rely on getting the type back that the function name promises which is a design problem.
I am unsure if I understand your point on the different indices for vertex and states. I added functions to the state space that should allow the user to not care about those indices.

> The core idea of state space is to be a "proxy" to the underlying graph, where the functions just forward under a different name that better matches the context of state spaces. Perhaps it is a bit misleading to use the naming `get_states` instead of `get_state_vertices`, etc. The benefit of this idea is that you can throw the underlying graph into the boost BGL library to compute useful stuff for the state space.

Sure, and that's great to have as an option! But I am sure you agree that a user should be able to get the info out they need without jumping through hoops.

> The other boilerplate I was referring to is in `bindings.cpp`: e.g. `insert_into_list` it looks unnessary given that one can fix the keep-alive chain through opaqueness as it is done in other cases already.

that boilerplate is necessary for when you need to return fluent/derived/static whatever types in a single list. To not rewrite it every time I now refactored it into a free function, prevents any more keep-alive patient-nurse switcheroos and saves time for other such cases in the future.

On a another note, the state space has more of those `get_xyz_state` functions that seemingly don't return states, but other intermediates. It might make sense to convert all those names wholesale before having a mix of names with different meanings.

drexlerd commented:
> > Regarding the API change: Notice that the vertex indices do not necessarily match the state indices. Now, if I want to access the successor states of the initial state, I first have to get the vertex index somehow. I don't see how adding these duplicate functions will make things easier. The current API will give you the index to the root vertex in the graph and you can easily access the successor vertices and their underlying states.
> 
> You can perhaps easily access the successor indices, but the name `state` is too overloaded at the moment. Users of the lib cannot rely on getting the type back that the function name promises which is a design problem. I am unsure if I understand your point on the different indices for vertex and states. I added functions to the state space that should allow the user to not care about those indices.
> 

A state space is generally does not contain all possible states, since it is computed in a forward search from the initial state. The state repository may generally contain states computed in a backward manner. Depending on whether the user performed some backward search from the goal, the state indices in the state space may begin with `n,n+1,n+2,...`. In this case, state indices will not match the vertices in the state space. Another example could be that a user manually constructs a state in the repository with index 0, that shifts the indices of states encountered during state space construction.

> > The core idea of state space is to be a "proxy" to the underlying graph, where the functions just forward under a different name that better matches the context of state spaces. Perhaps it is a bit misleading to use the naming `get_states` instead of `get_state_vertices`, etc. The benefit of this idea is that you can throw the underlying graph into the boost BGL library to compute useful stuff for the state space.
> 
> Sure, and that's great to have as an option! But I am sure you agree that a user should be able to get the info out they need without jumping through hoops.
> 

I disagree that user have to jump through hoops to access the information they need. I prefer an interface that is as compact as possible without loosing expressiveness. In the current design you can easily traverse through the state space with almost always a single `get_state` function call on the vertices that you get returned. The aim of the API should be to make it consistently one such call.

> > The other boilerplate I was referring to is in `bindings.cpp`: e.g. `insert_into_list` it looks unnessary given that one can fix the keep-alive chain through opaqueness as it is done in other cases already.
> 
> that boilerplate is necessary for when you need to return fluent/derived/static whatever types in a single list. To not rewrite it every time I now refactored it into a free function, prevents any more keep-alive patient-nurse switcheroos and saves time for other such cases in the future.
> 
> On a another note, the state space has more of those `get_xyz_state` functions that seemingly don't return states, but other intermediates. It might make sense to convert all those names wholesale before having a mix of names with different meanings.

I agree the current names are bad. They should better reflect that they are state vertices or state vertex indices.

maichmueller commented:
> > > Regarding the API change: Notice that the vertex indices do not necessarily match the state indices. Now, if I want to access the successor states of the initial state, I first have to get the vertex index somehow. I don't see how adding these duplicate functions will make things easier. The current API will give you the index to the root vertex in the graph and you can easily access the successor vertices and their underlying states.
> > 
> > 
> > You can perhaps easily access the successor indices, but the name `state` is too overloaded at the moment. Users of the lib cannot rely on getting the type back that the function name promises which is a design problem. I am unsure if I understand your point on the different indices for vertex and states. I added functions to the state space that should allow the user to not care about those indices.
> 
> A state space generally does not contain all possible states, since it is computed in a forward search from the initial state. The state repository may generally also contain states computed in a backward manner from the goal. Depending on whether the user performed some backward search from the goal, the state indices in the state space may begin with `n,n+1,n+2,...`. In this case, state indices will not match the vertices in the state space. Another example could be that a user manually constructs a state in the repository with index 0, that shifts the indices of states encountered during state space construction.
> 

That's good to know (and quite the change from the previous state space implementation). A little off-topic: is there a way then to force the statespace to compute all possible states at once? Or would they be generated on demand?

> > > The core idea of state space is to be a "proxy" to the underlying graph, where the functions just forward under a different name that better matches the context of state spaces. Perhaps it is a bit misleading to use the naming `get_states` instead of `get_state_vertices`, etc. The benefit of this idea is that you can throw the underlying graph into the boost BGL library to compute useful stuff for the state space.
> > 
> > 
> > Sure, and that's great to have as an option! But I am sure you agree that a user should be able to get the info out they need without jumping through hoops.
> 
> I disagree that user have to jump through hoops to access the information they need. I prefer an interface that is as compact as possible without loosing expressiveness. In the current design you can easily traverse through the state space with almost always a single `get_state` function call on the vertices that you get returned. The aim of the API should be to make it consistently one such call.
> 

ok, I don't see why a user should not be given the convenience to side-step the vertex if a state is all they want out of an index, but I will remove those defs if you feel strongly about it.

drexlerd commented:
> > > > Regarding the API change: Notice that the vertex indices do not necessarily match the state indices. Now, if I want to access the successor states of the initial state, I first have to get the vertex index somehow. I don't see how adding these duplicate functions will make things easier. The current API will give you the index to the root vertex in the graph and you can easily access the successor vertices and their underlying states.
> > > 
> > > 
> > > You can perhaps easily access the successor indices, but the name `state` is too overloaded at the moment. Users of the lib cannot rely on getting the type back that the function name promises which is a design problem. I am unsure if I understand your point on the different indices for vertex and states. I added functions to the state space that should allow the user to not care about those indices.
> > 
> > 
> > A state space generally does not contain all possible states, since it is computed in a forward search from the initial state. The state repository may generally also contain states computed in a backward manner from the goal. Depending on whether the user performed some backward search from the goal, the state indices in the state space may begin with `n,n+1,n+2,...`. In this case, state indices will not match the vertices in the state space. Another example could be that a user manually constructs a state in the repository with index 0, that shifts the indices of states encountered during state space construction.
> 
> That's good to know (and quite the change from the previous state space implementation). A little off-topic: is there a way then to force the statespace to compute all possible states at once? Or would they be generated on demand?
> 
This would require an additional backward search from the goal. Our successor generator computes forward applicable actions only. The current version and the old version compute equivalent sets of states.

Let's assume we would add the backward exploration as well. Then users can still break the indexing, by constructing unsolvable states in the state repository. 

> > > > The core idea of state space is to be a "proxy" to the underlying graph, where the functions just forward under a different name that better matches the context of state spaces. Perhaps it is a bit misleading to use the naming `get_states` instead of `get_state_vertices`, etc. The benefit of this idea is that you can throw the underlying graph into the boost BGL library to compute useful stuff for the state space.
> > > 
> > > 
> > > Sure, and that's great to have as an option! But I am sure you agree that a user should be able to get the info out they need without jumping through hoops.
> > 
> > 
> > I disagree that user have to jump through hoops to access the information they need. I prefer an interface that is as compact as possible without loosing expressiveness. In the current design you can easily traverse through the state space with almost always a single `get_state` function call on the vertices that you get returned. The aim of the API should be to make it consistently one such call.
> 
> ok, I don't see why a user should not be given the convenience to side-step the vertex if a state is all they want out of an index, but I will remove those defs if you feel strongly about it.

It is adding almost twice the amount of API, and inconsistencies with existing code such as the abstractions, which would also require a further inspection. It is too early to focus on such convenience functions that put even more workload on later refactorings.

maichmueller commented:
> This would require an additional backward search from the goal. Our successor generator computes forward applicable actions only. The current version and the old version compute equivalent sets of states.

> Let's assume we would add the backward exploration as well. Then users can still break the indexing, by constructing unsolvable states in the state repository.

I was wondering more in line with: For learning I need to sample out of all possible (valid) states in the state space. Whether this is currently possible.

> It is adding almost twice the amount of API, and inconsistencies with existing code such as the abstractions, which would also require a further inspection. It is too early to focus on such convenience functions that put even more workload on later refactorings.

> The only convenience functions that we currently add consistently is to retrieve a single element from a vector, to avoid the linear copying when going from c++ to python.

Alright, so remove all actual `get_..._state` functions, but keep the renaming to `get_...state_index` and `get_..._state_vertex`, right?

drexlerd commented:
> > This would require an additional backward search from the goal. Our successor generator computes forward applicable actions only. The current version and the old version compute equivalent sets of states.
> 
> > Let's assume we would add the backward exploration as well. Then users can still break the indexing, by constructing unsolvable states in the state repository.
> 
> I was wondering more in line with: For learning I need to sample out of all possible (valid) states in the state space. Whether this is currently possible.

Only from the forward reachable states. When your training set contains of several problems, it might be possible that one instance has an initial state that is not forward reachable in another instance. 

> 
> > It is adding almost twice the amount of API, and inconsistencies with existing code such as the abstractions, which would also require a further inspection. It is too early to focus on such convenience functions that put even more workload on later refactorings.
> 
> > The only convenience functions that we currently add consistently is to retrieve a single element from a vector, to avoid the linear copying when going from c++ to python.
> 
> Alright, so remove all actual `get_..._state` functions, but keep the renaming to `get_...state_index` and `get_..._state_vertex`, right?

Yes. I have not put much thought into a better naming. Perhaps you can come up with one. Perhaps adding suffixes `_state_vertex`, `_state_vertex_index` or `_vertex` and `_vertex_index` will make it. It is quite a lot of writing though. That is possibly the reasoning I decided to use those misleading names.

E.g. `get_initial_state_vertex_index` vs `get_initial_vertex_index`
Similarly: `get_goal_state_vertex_indices` vs `get_goal_vertex_indices`

The latter looks sufficient since there is just a single vertex type. Similarly for edges. This naming also matches the iterator names. This is just my current idea.

drexlerd commented:
I just checked the commits and it looks like we are not quite on the same page yet. You could remove lots of functions to reduce maintenance burden by using the method `get_vertex(Index)` (or `get_state_vertex(Index)` depending on what name you think is better) instead of adding `get_state` because `get_vertex(Index)` is already an O(1) operation.
The function `get_goal_states` should still retrieve the vertex indices but could be renamed more appropriately.

maichmueller commented:
You're too early ;) I only rebased yesterday onto the updated main

drexlerd commented:
The latest changes should resolve the API issues discussed in this pull request. The names now closely follow the naming convention used in general graphs and it is made explicit now that the indices refer to vertex indices and not state indices. I close the pull request for now.
